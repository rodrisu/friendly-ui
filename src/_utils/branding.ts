type Color = {
  [propName: string]: string
}

export const color: Color = {
  white: '#FFF',
  gray: '#DDD',
  lightGray: '#EDEDED',
  midnightGreen: '#054752',
  midnightGreenDetail: '#3D5C62',
  lightMidnightGreen: '#708C91',
  blue: '#00AFF5',
  darkBlue: '#008FC1',
  boostBlue: '#2ED1FF',
  green: '#5DD167',
  orange: '#F78B00',
  yellow: '#FFCA0C',
  red: '#F53F5B',
  lightRed: '#FDD8DE',
  facebookBrand: '#1877F2',
  vkBrand: '#2787F5',
  blaBlaCarDailyBrand: '#00D8A7',
  blaBlaCarDailyBrandText: '#30004B',
  tapHighlight: 'rgba(221, 221, 221, .4)', // gray, 40%
  lightBlue: '#ADECFE',
}

export const gradientColors = {
  blueGradient: `linear-gradient(80.19deg, ${color.boostBlue} 4.92%, ${color.blue} 95.98%)`,
}

type Font = {
  [key: string]: {
    size: string
    lineHeight: string
  }
}

type defaultBranding = {
  [key: string]: string
}

export const CDN_FONTS_BASE_URL = 'https://cdn.blablacar.com/kairos/assets/build/fonts'

export const font: Font = {
  s: {
    size: '13px',
    lineHeight: '16px',
  },
  base: {
    size: '16px',
    lineHeight: '20px',
  },
  m: {
    size: '18px',
    lineHeight: '20px',
  },
  l: {
    size: '22px',
    lineHeight: '24px',
  },
  xl: {
    size: '30px',
    lineHeight: String(Math.floor((32 / 30) * 100) / 100),
  },
  brand: {
    size: '56px',
    lineHeight: '56px',
  },
  xxl: {
    size: '82px',
    lineHeight: '82px',
  },
}

export const fontWeight: defaultBranding = {
  regular: '400',
  medium: '500',
}

export const space: defaultBranding = {
  none: '0',
  xs: '2px',
  s: '4px',
  m: '8px',
  l: '16px',
  xl: '24px',
  xxl: '48px',
}

export const horizontalSpace: defaultBranding = {
  inner: '8px',
  outer: '16px',
  global: '24px',
}

export const radius: defaultBranding = {
  none: '0',
  xs: '2px',
  s: '4px',
  m: '8px',
  l: '16px',
  xl: '24px',
}

// In milliseconds.
export enum TransitionDuration {
  BASE = 200,
  FAST = 500,
}

export const transition = {
  duration: {
    base: `${TransitionDuration.BASE}ms`,
    fast: `${TransitionDuration.FAST}ms`,
  },
  easing: {
    default: 'ease-in',
  },
  callbackDelay: 1000,
}

export const delay = {
  timeout: {
    base: '400ms',
  },
  interval: {
    base: '100ms',
  },
}

export const componentSizes = {
  timeWidth: '48px',
  buttonIconSize: '48px',
  bulletSize: '10px',
  bulletSizeSmall: '8px',
  bulletSizeMap: '18px',
  bulletSizeSearch: '18px',
  roadWidth: '4px',
  smallSectionWidth: '662px',
  largeSectionWidth: '1016px',
  modalTopPadding: '70px',
  searchOverlayWidth: '375px',
  headerHeight: {
    small: '56px',
    large: '72px',
  },
  headerWidth: {
    small: '100%',
    large: '1280px',
  },
  drawerWidth: {
    small: '100%',
    large: '350px',
  },
}

export const modalSize = {
  s: '400px',
  m: '662px',
  l: '928px',
}

export const shadow: defaultBranding = {
  default: '0 2px 4px rgba(0, 0, 0, .3)',
  icon: '0 0 2px rgba(0, 0, 0, .3)',
  card: '0 1pt 4pt rgba(0, 0, 0, 0.16), 0 2pt 8pt rgba(0, 0, 0, 0.08)',
  cardHover: '0 2pt 8pt rgba(0, 0, 0, 0.08), 0 2pt 16pt rgba(0, 0, 0, 0.08)',
  searchForm: '0 1pt 4pt rgba(0, 0, 0, 0.16), 0 2pt 8pt rgba(0, 0, 0, 0.08)',
  slideSection: '0 4px 4px rgba(0, 0, 0, .25)',
}

export const inputBorderSize = {
  default: '1px',
  focus: '3px',
}

const largeBreakPointThreshold = '800px'
const smallBreakPointThreshold = '799px'
export const responsiveBreakpoints = {
  small: smallBreakPointThreshold,
  isMediaLarge: `min-width: ${largeBreakPointThreshold}`,
  isMediaSmall: `max-width: ${smallBreakPointThreshold}`,
}

export const selectHeight = '52px'

export const pxToInteger = (size: string): number => parseInt(size.replace('px', ''), 10)

export const branding = {
  color,
  font,
  fontWeight,
  space,
  radius,
  shadow,
}
