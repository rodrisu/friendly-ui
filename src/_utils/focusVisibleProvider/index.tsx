import React, { createContext, ReactNode, useEffect, useState } from 'react'

import { KEYS, KEYS_TRIGGERING_KEYBOARD_NAVIGATION } from '../keycodes'

// A React hook based on: https://github.com/WICG/focus-visible

const pointerMoveEventList = [
  'mousedown',
  'mouseup',
  'pointermove',
  'pointerdown',
  'pointerup',
  'touchmove',
  'touchstart',
  'touchend',
]

const pointerDownEventList = ['mousedown', 'pointerdown', 'touchstart']

export const FOCUS_VISIBLE_CSS_CLASS = 'focus-visible'

export type FocusVisibleProviderProps = {
  children: ReactNode
  // Provide a global css selector for legacy components.
  setGlobalClassName?: boolean
}

export const FocusVisibleContext = createContext(false)

export const FocusVisibleProvider = ({
  children,
  setGlobalClassName,
}: FocusVisibleProviderProps) => {
  // When the provider first loads, assume the user is in pointer modality.
  const [hadKeyboardEvent, setHadKeyboardEvent] = useState(false)

  if (setGlobalClassName) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useEffect(() => {
      // Add a 'focus-visible' CSS class on the <body> element.
      document.body.classList.toggle(FOCUS_VISIBLE_CSS_CLASS, hadKeyboardEvent)
    }, [hadKeyboardEvent])
  }

  useEffect(() => {
    let lastClientX: Number
    let lastClientY: Number

    const onPointerDown = () => {
      setHadKeyboardEvent(false)
    }

    const onInitialPointerMove = () => {
      setHadKeyboardEvent(false)
    }

    const getLastMouseMove = (e: MouseEvent) => {
      // Ensure the mouse has actually moved (Safari)
      // https://transitory.technology/mouse-move-in-safari/
      if (lastClientX === e.clientX && lastClientY === e.clientY) {
        return
      }

      lastClientX = e.clientX
      lastClientY = e.clientY
    }

    const addInitialPointerMoveListeners = () => {
      document.addEventListener('mousemove', (event: MouseEvent) => {
        getLastMouseMove(event)
        // eslint-disable-next-line no-unused-expressions
        onInitialPointerMove
      })

      pointerMoveEventList.forEach(e => document.addEventListener(e, onInitialPointerMove))
    }

    const removeInitialPointerMoveListeners = () => {
      document.removeEventListener('mousemove', (event: MouseEvent) => {
        getLastMouseMove(event)
        // eslint-disable-next-line no-unused-expressions
        onInitialPointerMove
      })

      pointerMoveEventList.forEach(e => document.removeEventListener(e, onInitialPointerMove))
    }

    const onKeyDown = (e: KeyboardEvent) => {
      const element = e.target as HTMLInputElement
      const isTypingArea =
        element.tagName === 'TEXTAREA' || (element.tagName === 'INPUT' && element.type === 'text')
      // Remove Spacebar and Enter keys in case of text editing
      const keysList = isTypingArea
        ? KEYS_TRIGGERING_KEYBOARD_NAVIGATION.filter(
            key => key !== KEYS.SPACEBAR && key !== KEYS.ENTER,
          )
        : KEYS_TRIGGERING_KEYBOARD_NAVIGATION

      if (keysList.includes(e.key)) {
        setHadKeyboardEvent(true)
      }
    }

    // For some kinds of state, we are interested in changes at the global
    // scope only. Global pointer input and global key presses change
    // should affect the state at every scope.
    document.addEventListener('keydown', onKeyDown, true)
    pointerDownEventList.forEach(e => document.addEventListener(e, onPointerDown, true))
    addInitialPointerMoveListeners()

    return () => {
      document.removeEventListener('keydown', onKeyDown, true)
      pointerDownEventList.forEach(e => document.removeEventListener(e, onPointerDown, true))
      removeInitialPointerMoveListeners()
    }
  }, [])

  return (
    <FocusVisibleContext.Provider value={hadKeyboardEvent}>{children}</FocusVisibleContext.Provider>
  )
}
