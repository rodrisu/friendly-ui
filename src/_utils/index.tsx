import React, { Fragment } from 'react'
import { canUseDOM } from 'exenv'

const DEFAULT_CLASS_PREFIX = 'kirk'

export const prefix = (modifiers: Classcat.ClassObject, baseClass?: string): string => {
  const mods = Object.keys(modifiers).filter(elem => modifiers[elem])

  let classPrefix = DEFAULT_CLASS_PREFIX

  if (baseClass != null) {
    classPrefix = `${classPrefix}-${baseClass}`
  }

  return mods.map(modifier => `${classPrefix}-${modifier}`).join(' ')
}

export const isTouchEventsAvailable = () =>
  canUseDOM && 'ontouchstart' in window && 'ontouchend' in window

export const replaceNewLineWithBR = (str: string): React.ReactNode =>
  str
    .split('\n')
    .map(line => <Fragment>{line}</Fragment>)
    .reduce((acc, curr) => (
      <Fragment>
        {acc}
        <br />
        {curr}
      </Fragment>
    ))
