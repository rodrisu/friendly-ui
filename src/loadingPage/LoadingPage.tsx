import React from 'react'

import { TheVoice } from '../theVoice'
import { StyledLoadingPage } from './LoadingPage.style'

export type LoadingPageProps = Readonly<{
  icon?: React.ElementType
  iconColor?: string
  iconBackground?: string
  title?: string
}>

export const LoadingPage = ({
  icon: DisplayedIcon,
  iconColor,
  iconBackground,
  title,
}: LoadingPageProps) => (
  <StyledLoadingPage>
    <div className="content-container">
      {DisplayedIcon && (
        <div className="logo" style={{ backgroundColor: iconBackground }}>
          <DisplayedIcon iconColor={iconColor} />
        </div>
      )}
      {title && <TheVoice>{title}</TheVoice>}
    </div>
  </StyledLoadingPage>
)
