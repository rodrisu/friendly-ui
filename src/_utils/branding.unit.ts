import { pxToInteger } from './branding'

it('should convert a string px value to integer', () => {
  expect(pxToInteger('10px')).toEqual(10)
})
