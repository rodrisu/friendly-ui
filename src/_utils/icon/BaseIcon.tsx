import React from 'react'
import cc from 'classcat'
import isEmpty from 'lodash.isempty'

import { Badge } from '../../badge'
import { color } from '../branding'

export type Icon = Readonly<{
  size?: number | string
  className?: string
  iconClassName?: string
  title?: string
  iconColor?: string
  badgeAriaLabel?: string
  badgeContent?: string
  isDisabled?: boolean
}>

export type IconProps = Icon &
  Readonly<{
    children: JSX.Element
    viewBox?: string
  }>

export const BaseIconDefaultProps = {
  className: '',
  iconClassName: '',
  iconColor: color.lightMidnightGreen,
  size: 24,
  title: '',
  badgeAriaLabel: '',
  badgeContent: '',
  isDisabled: false,
}

export const BaseIcon = ({
  className,
  iconClassName,
  size,
  title,
  children,
  viewBox = '0 0 24 24',
  badgeContent,
  badgeAriaLabel,
}: IconProps) => {
  const iconClasses = ['kirk-icon', iconClassName]
  if (!badgeContent) {
    // When there is no badge, the icon is not wrapped and is the root element.
    // Therefore className classes needs to be added to the icon itself.
    iconClasses.push(className)
  }

  const icon = (
    <svg
      viewBox={viewBox}
      xmlns="http://www.w3.org/2000/svg"
      className={cc(iconClasses)}
      width={size}
      height={size}
      aria-hidden={isEmpty(title)}
    >
      {title && <title>{title}</title>}
      {children}
    </svg>
  )

  if (badgeContent) {
    return (
      <div className={cc(['kirk-icon-wrapper', className])}>
        {icon}
        <Badge className="kirk-icon-badge" aria-label={badgeAriaLabel}>
          {badgeContent}
        </Badge>
      </div>
    )
  }

  return icon
}
