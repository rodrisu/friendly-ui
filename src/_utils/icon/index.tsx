import styled from 'styled-components'

import { color, shadow } from '../branding'
import { BaseIcon } from './BaseIcon'

const StyledBaseIcon = styled(BaseIcon)`
  fill: ${props => props.iconColor ?? color.lightMidnightGreen};

  &,
  & rect,
  & circle,
  & line,
  & polyline,
  & path,
  & g {
    stroke: ${props => (props.isDisabled ? color.gray : '')};
  }

  &.kirk-icon-wrapper {
    display: inline-block;
    position: relative;
  }

  &.kirk-icon-shadowed {
    border-radius: 50%;
    box-shadow: ${shadow.icon};
  }

  & .kirk-icon-badge {
    position: absolute;
    bottom: 12px;
    left: 12px;
  }
`

export * from './BaseIcon'

export { StyledBaseIcon as BaseIcon }
