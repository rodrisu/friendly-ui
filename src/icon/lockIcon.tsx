import React from 'react'

import { BaseIcon, BaseIconDefaultProps, Icon } from '../_utils/icon'

export const LockIcon = (props: Icon) => (
  <BaseIcon {...props}>
    <path
      fill={props.iconColor}
      fillRule="evenodd"
      d="M7.5 9.5v-3a4.5 4.5 0 019 0v3H18a1 1 0 011 1V21a1 1 0 01-1 1H6a1 1 0 01-1-1V10.5a1 1 0 011-1h1.5zm1 0h7v-3a3.5 3.5 0 00-7 0v3zm-2.5 1V21h12V10.5H6zm5.5 6.45a2.5 2.5 0 111 0v1.55a.5.5 0 11-1 0v-1.55zM12 16a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
    />
  </BaseIcon>
)

LockIcon.defaultProps = BaseIconDefaultProps
