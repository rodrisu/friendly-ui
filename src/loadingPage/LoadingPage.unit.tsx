import React from 'react'
import renderer from 'react-test-renderer'

import { color } from '../_utils/branding'
import { LockIcon } from '../icon/lockIcon'
import { LoadingPage } from './LoadingPage'

it('Should have the correct attributes and text when all props.', () => {
  const loadingPage = renderer
    .create(
      <LoadingPage
        icon={LockIcon}
        iconColor={color.white}
        iconBackground={color.green}
        title="Entering a secure page"
      />,
    )
    .toJSON()

  expect(loadingPage).toMatchSnapshot()
})

it('Should have the correct attributes and text when no icon.', () => {
  const loadingPage = renderer.create(<LoadingPage title="Entering a secure page" />).toJSON()

  expect(loadingPage).toMatchSnapshot()
})
