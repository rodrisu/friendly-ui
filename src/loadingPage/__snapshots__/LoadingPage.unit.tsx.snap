// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have the correct attributes and text when all props. 1`] = `
.c1 {
  fill: #FFF;
}

.c1.kirk-icon-wrapper {
  display: inline-block;
  position: relative;
}

.c1.kirk-icon-shadowed {
  border-radius: 50%;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
}

.c1 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

.c2 {
  white-space: pre-line;
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
  font-weight: 500;
  --space-bottom: 32px;
  --font-size-desktop: 40px;
  margin: 0;
  padding: 24px 0 var(--space-bottom);
  padding-left: 24px !important;
  padding-right: 24px !important;
  margin-left: 0;
  margin-right: 0;
}

.c0 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFF;
}

.c0 .content-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .content-container .logo {
  border-radius: 100%;
  width: 48px;
  height: 48px;
}

.c0 .content-container .logo svg {
  margin: 8px 9px;
  width: 30px;
  height: 30px;
}

@media (min-width:800px) {
  .c2 {
    text-align: center;
    font-size: var(--font-size-desktop);
  }
}

<div
  className="c0"
>
  <div
    className="content-container"
  >
    <div
      className="logo"
      style={
        Object {
          "backgroundColor": "#5DD167",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="kirk-icon c1"
        height={24}
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <path
          d="M7.5 9.5v-3a4.5 4.5 0 019 0v3H18a1 1 0 011 1V21a1 1 0 01-1 1H6a1 1 0 01-1-1V10.5a1 1 0 011-1h1.5zm1 0h7v-3a3.5 3.5 0 00-7 0v3zm-2.5 1V21h12V10.5H6zm5.5 6.45a2.5 2.5 0 111 0v1.55a.5.5 0 11-1 0v-1.55zM12 16a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
          fill="#FFF"
          fillRule="evenodd"
        />
      </svg>
    </div>
    <h1
      className="c2"
    >
      Entering a secure page
    </h1>
  </div>
</div>
`;

exports[`Should have the correct attributes and text when no icon. 1`] = `
.c1 {
  white-space: pre-line;
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
  font-weight: 500;
  --space-bottom: 32px;
  --font-size-desktop: 40px;
  margin: 0;
  padding: 24px 0 var(--space-bottom);
  padding-left: 24px !important;
  padding-right: 24px !important;
  margin-left: 0;
  margin-right: 0;
}

.c0 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFF;
}

.c0 .content-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .content-container .logo {
  border-radius: 100%;
  width: 48px;
  height: 48px;
}

.c0 .content-container .logo svg {
  margin: 8px 9px;
  width: 30px;
  height: 30px;
}

@media (min-width:800px) {
  .c1 {
    text-align: center;
    font-size: var(--font-size-desktop);
  }
}

<div
  className="c0"
>
  <div
    className="content-container"
  >
    <h1
      className="c1"
    >
      Entering a secure page
    </h1>
  </div>
</div>
`;
